mypy/typeanal.py

```diff
@@ -712,13 +739,22 @@ def visit_type_list(self, t: TypeList) -> None:
         self.fail('Invalid type', t)
 
     def visit_type_var(self, t: TypeVarType) -> None:
-        pass
+        if t.upper_bound:
+            t.upper_bound.accept(self)
+        if t.values:
+            for v in t.values:
+                v.accept(self)
 
     def visit_partial_type(self, t: PartialType) -> None:
         pass
 
     def visit_type_type(self, t: TypeType) -> None:
-        pass
+        t.item.accept(self)
+
+    def visit_forwardref_type(self, t: ForwardRef) -> None:
+        self.indicator['forward'] = True
+        if isinstance(t.link, UnboundType):
+            t.link = self.sem.anal_type(t.link, third_pass=True)
```

Comment:
This is kind of hacky -- we call the second pass of semantic analysis from the third pass of type analysis, without having all the normal context for the second pass set up. This causes some errors:

```py
from typing import List, TypeVar

T = TypeVar('T')

def f(x: T) -> None:
    y: A[T]  # "Invalid type "t.T"

A = List[T]
```

Another example that doesn't work because of this issue:

```py
    from typing import List, TypeVar

    def f() -> None:
        X = List[int]
        x: A[X]  # "Invalid type X?"

    T = TypeVar('T')
    A = List[T]
```

Not sure what's the best way to approach this. We perhaps don't want to set up the full context for second pass during the third pass, as this can add a lot of extra complexity. A simpler workaround would be to not support forward references with type arguments, and give an error such as `Forward reference to "A" not supported`. This at least would fix the current error messages, which are confusing.

Also, I'd prefer if we didn't depend on `mypy.semanal` here. A better approach would be to remove the cyclic import to `mypy.semanal` and call `TypeAnalyser` directly here. `TypeAnalyserPass3` would take the lookup functions etc. as arguments instead of getting them indirectly from `SemanticAnalyzer`. The rationale is that this makes it explicit what context the third pass depends on, and it will make the code clearer. Also, we won't depend on the internals of `SemanticAnalyzer`.

Reply:
Yes, this is one of the hard parts. I will try to see what is the minimal context needed, basically we just need ``lookup`` to work in third pass. If it will be too hard, I will just prohibit things like ``F[<something>]``, where ``F`` is a forward reference, with a clearer error message.